### PR TITLE
Alter exception handling to utilize response data

### PR DESCRIPTION
**Pull request by bachya**
# Proposed Changes

Following up on https://github.com/frenck/python-demetriek/pull/275#issuecomment-1309103412. The current HTTP exception handling swallows important response data, preventing implementors from more nuance in their applications. This PR alters things to include the response data (when appropriate) in the raised exception.

As an example, this is the current response when attempting to send a non-critical notification to a device with screensaver mode active:

```
Traceback (most recent call last):
  File "/Users/bachya/Git/python-demetriek/src/demetriek/device.py", line 77, in _request
    response = await self.session.request(
  File "/Users/bachya/.virtualenvs/python-demetriek/lib/python3.10/site-packages/aiohttp/client.py", line 643, in _request
    resp.raise_for_status()
  File "/Users/bachya/.virtualenvs/python-demetriek/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1005, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('https://192.168.1.100:4343/api/v2/device/notifications')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bachya/Git/python-demetriek/examples/example.py", line 58, in <module>
    asyncio.run(main())
  File "/Users/bachya/.pyenv/versions/3.10.5/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/bachya/.pyenv/versions/3.10.5/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/bachya/Git/python-demetriek/examples/example.py", line 54, in main
    await lametric.notify(notification=notification)
  File "/Users/bachya/Git/python-demetriek/src/demetriek/device.py", line 243, in notify
    response = await self._request(
  File "/Users/bachya/.virtualenvs/python-demetriek/lib/python3.10/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
  File "/Users/bachya/Git/python-demetriek/src/demetriek/device.py", line 101, in _request
    raise LaMetricError(
demetriek.exceptions.LaMetricError: Error occurred while connecting to the LaMetric device at 192.168.1.100
```

With this PR, the exception now includes more useful information:

```
Traceback (most recent call last):
  File "/Users/bachya/Git/python-demetriek/src/demetriek/device.py", line 92, in _request
    response.raise_for_status()
  File "/Users/bachya/.virtualenvs/python-demetriek/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1005, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('https://192.168.1.100:4343/api/v2/device/notifications')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bachya/Git/python-demetriek/examples/example.py", line 58, in <module>
    asyncio.run(main())
  File "/Users/bachya/.pyenv/versions/3.10.5/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/bachya/.pyenv/versions/3.10.5/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/bachya/Git/python-demetriek/examples/example.py", line 54, in main
    await lametric.notify(notification=notification)
  File "/Users/bachya/Git/python-demetriek/src/demetriek/device.py", line 241, in notify
    response = await self._request(
  File "/Users/bachya/.virtualenvs/python-demetriek/lib/python3.10/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
  File "/Users/bachya/Git/python-demetriek/src/demetriek/device.py", line 94, in _request
    raise_on_data_error(self.host, response_data, exception)
  File "/Users/bachya/Git/python-demetriek/src/demetriek/exceptions.py", line 47, in raise_on_data_error
    raise LaMetricError(
demetriek.exceptions.LaMetricError: Error occurred while communicating with the LaMetric device at 192.168.1.100: Only notifications with priority 'critical' are allowed in current mode
```

Once the overall idea is approved, I'll add tests and make sure CI passes.

## Related Issues

> ([Github link][autolink-references] to related issues or pull requests)

[autolink-references]: https://help.github.com/articles/autolinked-references-and-urls/

### CODE REVIEW

**Comment by Anto79-ops** [comment]
I wonder if its possible to create a (binary) sensor for the clock that reports if its in screensaver mode or not (on-off), that way, when sending a message, a condition can be created so that it won't send non-critical messages to the clock if screen-save mode is activated.

For example, if you are not home and you want to send a message to a clock. you have no way of knowing if its in screen-saver mode.

**Comment by Anto79-ops** [comment]
also, I wanted to mention in the clocks latest firmware update to 2.3.1 on Nov 17, 2022

"New Local API now allows you to change device mode...."

I'm not exactly sure what the above means, if it means you can now wake the clock up via API?

**Comment by frenck** [comment]
> also, I wanted to mention in the clocks latest firmware update to 2.3.1 on Nov 17, 2022

> "New Local API now allows you to change device mode...."

Not related to this PR. Let keep things on topic. I'm aware of the changes, I'm in contact with LaMetric.

**Comment by bachya** [comment]
> I wonder if its possible to create a (binary) sensor for the clock that reports if its in screensaver mode or not (on-off), that way, when sending a message, a condition can be created so that it won't send non-critical messages to the clock if screen-save mode is activated.
> 
> 
> 
> For example, if you are not home and you want to send a message to a clock. you have no way of knowing if its in screen-saver mode.
> 
> 

Would be separate from this PR, but I think creating a "screensaver enabled?" binary sensor is very reasonable. 👍🏻

**Comment by frenck** [comment]
> Would be separate from this PR, but I think creating a "screensaver enabled?" binary sensor is very reasonable. 👍🏻

That is not a binary sensor. But also offtopic.

**Comment by bachya** [comment]
Still active.

**Comment by bachya** [comment]
Still active.

File: src/demetriek/exceptions.py [hunk 1]
```diff
@@ -15,3 +20,28 @@ class LaMetricAuthenticationError(LaMetricError):
 
 class LaMetricConnectionTimeoutError(LaMetricConnectionError):
     """LaMetric connection Timeout exception."""
+
+
+def raise_on_data_error(
```
Comment (frenck) [inline src/demetriek/exceptions.py:25]: We don't need to extract this, we can just do that inline (as it is only used once).

File: src/demetriek/device.py [hunk 2]
```diff
@@ -80,32 +81,35 @@ async def _request(
...
             if "application/json" not in content_type:
                 raise LaMetricError(response.status, {"message": await response.text()})
-            return await response.json()
+
+            response_data: dict[str, Any] = await response.json()
+
+            try:
+                response.raise_for_status()
+            except aiohttp.ClientResponseError as exception:
+                if exception.status in (401, 403):
```
Comment (frenck) [inline src/demetriek/device.py:96]: If we get here, we already did an `response.json()` call above. Do  401/403 responses contain JSON payloads?

**Review by frenck — CHANGES_REQUESTED** [review]

File: src/demetriek/device.py (hunk 2 above)
Comment (bachya) [reply src/demetriek/device.py:95]: Yes—in the case mentioned in the PR description, the `Only notifications with priority 'critical' are allowed in current mode` message is returned as part of the response payload along with a `HTTP 401`.

Comment (frenck) [reply src/demetriek/device.py:95]: OK 👍

**Comment by Anto79-ops** [comment]
not stale :)

**Pull request closed without merging**